datapath/test: Do not SNAT for WORLD_ID and enable BPF masquerading by default in CI

brb opened the pull request:
See commit msgs.

Depends on #11572 (let's merge this PR, and then close #11572).

Partially_fix #11442

brb committed:
test: Enable BPF masquerading

Enable BPF masquerading for the {{4.19,net-next},no-kube-proxy} CI jobs.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Disable BPF masq for K8sKubeProxyFreeMatrix

Disable the BPF masq in the vxlan tests until PublicInterfaceName is
decluttered. The communication between pod and remote node has to be
SNAT'd in the case of vxlan, which is currently not feasible, as
bpf_netdev is loaded only on PublicInterfaceName.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Do not set devices in DatapathConfiguration BPF masq tests

As we enabled the BPF masq by default, the correct devices (default and
private) are already set by DeployCilium().

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Set single --device in K8sUpdates test

Cilium < v1.8 doesn't support multi-dev, so installing e.g. v1.7 with
multiple devices (set by default in overwriteHelmOptions()) will crash
cilium-agent.

Therefore, in the upgrade test use a single device.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Get rid of dedicated BPF masq test cases

As BPF masq is enabled by default, we don't need a dedicated deployments
(and test cases) to test the feature.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Do not SNAT if dst id is WORLD_ID

Unfortunately, we cannot SNAT a packet from a local endpoint only if
dst sec id is WORLD_ID. The problem with this is that in the case of
an FQDN policy, a world destination can get its own sec id, which makes
a packet destined to such target to bypass the check, and bypass the
SNAT.

To fix this, we SNAT a packet from a local endpoint if dst is neither
from native-routing-cidr nor REMOTE_HOST_ID.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Set ipv4-alloc-range if native-routing-cidr is not available

Since we changed the SNAT logic in the previous commit, we need to set
IPV4_SNAT_EXCLUSION_DST_CIDR (prev. IPV4_NATIVE_ROUTING_CIDR) in any
case, otherwise traffic between pods on different nodes will be SNAT'd.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

qmonnet commented on test/helpers/kubectl.go:
```diff
@@ -1681,13 +1681,26 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
+			// Enable BPF masquerading
+			defaultIface, err := kub.GetDefaultIface()
+			if err != nil {
+				return err
```
Should this log an error message? Asking because tests below are removed in another commit, so it might be more difficult to troubleshoot if for some reason we fail to get the default interface?

``` diff
-		defaultIface, err = kubectl.GetDefaultIface()
-		Expect(err).Should(BeNil(), "Cannot determine default iface")
```

brb replied:
I think we should not put assertions into helper functions, and assert only in consumers of those helpers. E.g. in this case if the default iface cannot be determined, then assertions in `DeployCiliumAndDNS()` will fail with the msg saying something similar to "cannot deploy cilium because default iface cannot be determined".

qmonnet replied:
Ok, fair enough :+1:

qmonnet commented on bpf/lib/nodeport.h:
```diff
@@ -795,9 +798,14 @@ static __always_inline bool nodeport_nat_ipv4_needed(struct __ctx_buff *ctx,
+					return false;
+#endif /* ENCAP_IFINDEX */
+
+				return true;
```
Nit: Could we simplify:
``` c
        if (info->sec_label == REMOTE_NODE_ID)
#ifdef ENCAP_IFINDEX
                return true;
#else
                return false;
#endif
        return true;
```
Into:
``` c
#ifndef ENCAP_IFINDEX
        if (info->sec_label == REMOTE_NODE_ID)
                return false;
#endif
        return true;
```

brb committed:
datapath: Fix removal of bpf_netdev

The name of a bpf_netdev object file which is attached to a native
device has changed its name - from "bpf_netdev.o" to
"bpf_netdev_${NATIVE_DEV_IFACE}.o". E.g.:

$ sudo tc filter show dev enp0s8 ingress
filter protocol all pref 1 bpf chain 0
filter protocol all pref 1 bpf chain 0 handle 0x1 bpf_netdev_enp0s8.o:[from-netdev] direct-action not_in_hw id 982 tag feca2ca7f6f80c7e jited

The change of the name broke the removal of bpf_netdev from previously
used native devs.

Fixes: a695f532d06 ("Endpoint for host")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann reviewed (approved):
as per discussion offline
